File: leetcode/DP/322.Coin-Change/322.Coin-Change.go
package leetcode

/* 322. 零钱兑换
给你一个整数数组 coins ，表示不同面额的硬币；以及一个整数 amount ，
表示总金额。
计算并返回可以凑成总金额所需的 最少的硬币个数 。
如果没有任何一种硬币组合能组成总金额，返回 -1 。
你可以认为每种硬币的数量是无限的。

示例 1：
输入：coins = [1, 2, 5], amount = 11
输出：3 
解释：11 = 5 + 5 + 1

示例 2：
输入：coins = [2], amount = 3
输出：-1

示例 3：
输入：coins = [1], amount = 0
输出：0

*/


// 参考算法模板 和其他 DP 不太一样，i 表示钱或者容量
func coinChange(coins []int, amount int) int {
	// 状态 dp[i] 表示金额为i时,组成的最小硬币个数
	// 推导 dp[i] = min(dp[i-1], dp[i-2], dp[i-5])+1, 前提i-coins[j] > 0
	// 初始化为最大值 dp[i]=amount+1
	// 返回值 dp[n] or dp[n] > amount => -1
	dp := make([]int, amount+1)
	for i := 0; i <= amount; i++ {
		dp[i] = amount+1
	}
	dp[0] = 0
	for i := 1; i <= amount; i++ {
		for j := 0; j < len(coins); j++ {
			if i - coins[j] >= 0 {
				dp[i] = min(dp[i], dp[i-coins[j]]+1)
			}
		}
	}
	if dp[amount] > amount {
		return -1
	}
	return dp[amount]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

// 参考 halfrost
func coinChange_v1(coins []int, amount int) int {
	dp := make([]int, amount+1)
	dp[0] = 0 // base case
	for i := 1; i < len(dp); i++ { // 初始化dp数组为最大值 dp[i]=amount+1
		dp[i] = amount + 1
	}
	for i := 1; i <= amount; i++ { // 外层 for 循环在遍历所有状态的所有取值
		for j := 0; j < len(coins); j++ { // 内层for循环再求所有选择的最小值
			if coins[j] <= i { // 前提i-coins[j] > 0, 不满足的情况直接跳过
				dp[i] = min(dp[i], dp[i-coins[j]]+1) // 状态方程 选择最优子结构
			}
		}
	}
	if dp[amount] > amount { // 没有任何一种硬币组合能组成总金额
		return -1
	}
	return dp[amount] 
}

/* 
解题思路
给出一些硬币和一个总数，问组成这个总数的硬币数最少是多少个？
这一题是经典的硬币问题，利用 DP 求解。不过这一题的测试用例有一个很大的值，
这样开 DP 数组会比较浪费空间。例如 [1,1000000000,500000] 有这样的硬币种类，
要求组成 2389412493027523 这样的总数。
那么按照下面的解题方法，数组会开的很大，
非常浪费空间。这个时候用 DFS 解题会节约一点空间。
*/

File: leetcode-cpp/322.Coin-Change/322.Coin-Change.cpp
/* 322. 零钱兑换
给你一个整数数组 coins ，表示不同面额的硬币；
以及一个整数 amount ，表示总金额。

计算并返回可以凑成总金额所需的 最少的硬币个数 。
如果没有任何一种硬币组合能组成总金额，返回 -1 。

你可以认为每种硬币的数量是无限的。

示例 1：

输入：coins = [1, 2, 5], amount = 11
输出：3 
解释：11 = 5 + 5 + 1
示例 2：

输入：coins = [2], amount = 3
输出：-1
示例 3：

输入：coins = [1], amount = 0
输出：0
 

提示：

1 <= coins.length <= 12
1 <= coins[i] <= 2^31 - 1
0 <= amount <= 10^4
*/

// 参考代码随想录
// 遍历方式:遍历物品放在外循环，遍历背包放在内循环
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        for (int i = 0; i < coins.size(); i++) { // 遍历物品
            for (int j = coins[i]; j <= amount; j++) { // 遍历背包
                if (dp[j - coins[i]] != INT_MAX) { // 如果dp[j-coins[i]]是初始值则跳过
                    dp[j] = min(dp[j - coins[i]] + 1, dp[j]);
                }
            }
        }
        if (dp[amount] == INT_MAX) {
            return -1;
        }
        return dp[amount];
    }
};

// 遍历方式:遍历背包放在外循环，遍历物品放在内循环
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) { // 遍历背包
            for (int j = 0; j < coins.size(); j++) { // 遍历物品
                if (i - coins[j] >= 0 && dp[i-coins[j]] != INT_MAX) { // 如果背包容量不够且dp[j-coins[i]]是初始值则跳过
                    dp[i] = min(dp[i - coins[j]] + 1, dp[i]);
                }
            }
        }
        if (dp[amount] == INT_MAX) {
            return -1;
        }
        return dp[amount];
    }
};

/**
 * @brief 
 * 思路
 * 
在动态规划：518.零钱兑换II 中我们已经兑换一次零钱了，这次又要兑换，套路不一样！

题目中说每种硬币的数量是无限的，可以看出是典型的完全背包问题。

动规五部曲分析如下：

1. 确定dp数组以及下标的含义
dp[j]：凑足总额为j所需钱币的最少个数为dp[j]

2. 确定递推公式
得到dp[j]（考虑coins[i]），只有一个来源，
dp[j - coins[i]]（没有考虑coins[i]）。

凑足总额为j - coins[i]的最少个数为dp[j - coins[i]]，
那么只需要加上一个钱币coins[i]即dp[j - coins[i]] + 1
就是dp[j]（考虑coins[i]）

所以dp[j] 要取所有 dp[j - coins[i]] + 1 中最小的。

递推公式：dp[j] = min(dp[j - coins[i]] + 1, dp[j]);

3. dp数组如何初始化
首先凑足总金额为0所需钱币的个数一定是0，那么dp[0] = 0;

其他下标对应的数值呢？

考虑到递推公式的特性，dp[j]必须初始化为一个最大的数，
否则就会在min(dp[j - coins[i]] + 1, dp[j])比较的过程中被初始值覆盖。

所以下标非0的元素都是应该是最大值。

代码如下：

vector<int> dp(amount + 1, INT_MAX);
dp[0] = 0;

4. 确定遍历顺序
本题求钱币最小个数，那么钱币有顺序和没有顺序都可以，
都不影响钱币的最小个数。

所以本题并不强调集合是组合还是排列。

如果求组合数就是外层for循环遍历物品，内层for遍历背包。

如果求排列数就是外层for遍历背包，内层for循环遍历物品。

在动态规划专题我们讲过了求组合数是动态规划：518.零钱兑换II ，
求排列数是动态规划：377. 组合总和 Ⅳ 。

所以本题的两个for循环的关系是：外层for循环遍历物品，内层for遍历背包
或者外层for遍历背包，内层for循环遍历物品都是可以的！

那么我采用coins放在外循环，target在内循环的方式。

本题钱币数量可以无限使用，那么是完全背包。所以遍历的内循环是正序

综上所述，遍历顺序为：
coins（物品）放在外循环，target（背包）在内循环。且内循环正序。

举例推导dp数组
以输入：coins = [1, 2, 5], amount = 5为例

dp[amount]为最终结果。

#C++代码
 */